storage/key_index.go

```diff
@@ -0,0 +1,205 @@
+// removed, the keyIndex Should be removed.
+
+// For example:
+// compact(2) on the previous example
+// generations:
+//    {empty}
+//    {4, 5(t)}
+//    {2, 3(t)}
+//
+// compact(4)
+// generations:
+//    {empty}
+//    {4, 5(t)}
+//
+// compact(5):
+// generations:
+//    {empty}
+//    {5(t)}
+//
+// compact(6):
+// generations:
+//    {empty} -> key SHOULD be removed.
+type keyIndex struct {
+	key         []byte
+	index       uint64
+	generations []generation
+}
+
+// put puts an index to the keyIndex.
+func (ki *keyIndex) put(index uint64) {
+	if index < ki.index {
+		log.Panicf("store.keyindex: put with unexpected smaller index [%d / %d]", index, ki.index)
+	}
+	if len(ki.generations) == 0 {
+		ki.generations = append(ki.generations, generation{})
+	}
+	g := &ki.generations[len(ki.generations)-1]
+	g.cont = append(g.cont, index)
+	g.ver++
+	ki.index = index
+}
+
+// tombstone puts an index, pointing to a tombstone, to the keyIndex.
+// It also creates a new empty generation in the keyIndex.
+func (ki *keyIndex) tombstone(index uint64) {
+	if ki.isEmpty() {
+		log.Panicf("store.keyindex: unexpected tombstone on empty keyIndex %s", string(ki.key))
+	}
+	ki.put(index)
+	ki.generations = append(ki.generations, generation{})
+}
+
+// get gets the index of thk that satisfies the given atIndex.
+// Index must be lower or equal to the given atIndex.
+func (ki *keyIndex) get(atIndex uint64) (index uint64, err error) {
+	if ki.isEmpty() {
+		log.Panicf("store.keyindex: unexpected get on empty keyIndex %s", string(ki.key))
+	}
+	g := ki.findGeneration(atIndex)
+	if g.isEmpty() {
+		return 0, ErrIndexNotFound
+	}
+
+	f := func(index, ver uint64) bool {
+		if index <= atIndex {
+			return false
+		}
+		return true
+	}
+
+	_, n := g.walk(f)
+	if n != -1 {
+		return g.cont[n], nil
+	}
+	return 0, ErrIndexNotFound
+}
+
+// compact compacts a keyIndex by removing the versions with smaller or equal
+// index than the given atIndex except the largest one.
+// If a generation becomes empty during compaction, it will be removed.
+func (ki *keyIndex) compact(atIndex uint64, available map[uint64]struct{}) {
+	if ki.isEmpty() {
+		log.Panic("store.keyindex: unexpected compact on empty keyIndex %s", string(ki.key))
+	}
+	// walk until reaching the first content that has an index smaller or equal to
+	// the atIndex.
+	// add all the reached indexes into available map.
+	f := func(index, _ uint64) bool {
+		available[index] = struct{}{}
+		if index <= atIndex {
+			return false
+		}
+		return true
+	}
+
+	g := ki.findGeneration(atIndex)
+	i := len(ki.generations) - 1
+	for i >= 0 {
+		wg := &ki.generations[i]
+		if wg == g {
+			break
+		}
+		wg.walk(f)
+		i--
+	}
+
+	_, n := g.walk(f)
+
+	// remove the previous contents.
+	if n != -1 {
+		g.cont = g.cont[n:]
+	}
+	// remove the previous generations.
+	ki.generations = ki.generations[i:]
+
+	return
+}
+
+func (ki *keyIndex) isEmpty() bool {
+	return len(ki.generations) == 1 && ki.generations[0].isEmpty()
```

Comment:
so `{empty}` is counted as empty key index and ``(nothing) should never appear as a key index. `isEmpty` is used to check whether it is empty key index.
It makes me confused because we init it using `&keyIndex{index: index}`, which have empty generations. But it works well because we put something into it immediately. Maybe add a `newKeyIndex` function to make it more clear? Just for discussion, this is a trivial code improvement and there is no need to make any change here.